crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -0,0 +1,507 @@
+use crate::{
+    common::{
+        gas_price_algorithm::SharedGasPriceAlgo,
+        l2_block_source::L2BlockSource,
+        updater_metadata::UpdaterMetadata,
+        utils::BlockInfo,
+    },
+    ports::MetadataStorage,
+    v0::metadata::V0Metadata,
+    v1::{
+        algorithm::SharedV1Algorithm,
+        da_source_service::{
+            service::{
+                DaBlockCostsSource,
+                DaSourceService,
+                SharedState as DaSharedState,
+            },
+            DaBlockCosts,
+        },
+        metadata::{
+            v1_algorithm_from_metadata,
+            V1AlgorithmConfig,
+            V1Metadata,
+        },
+    },
+};
+use anyhow::anyhow;
+use async_trait::async_trait;
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    StateWatcher,
+};
+use fuel_gas_price_algorithm::{
+    v0::AlgorithmUpdaterV0,
+    v1::{
+        AlgorithmUpdaterV1,
+        AlgorithmV1,
+    },
+};
+use futures::FutureExt;
+use std::num::NonZeroU64;
+use tokio::sync::broadcast::Receiver;
+
+/// The service that updates the gas price algorithm.
+pub struct GasPriceServiceV1<L2, Metadata, DA>
+where
+    DA: DaBlockCostsSource,
+{
+    /// The algorithm that can be used in the next block
+    shared_algo: SharedV1Algorithm,
+    /// The L2 block source
+    l2_block_source: L2,
+    /// The metadata storage
+    metadata_storage: Metadata,
+    /// The algorithm updater
+    algorithm_updater: AlgorithmUpdaterV1,
+    /// the da source adapter handle
+    da_source_adapter_handle: DaSourceService<DA>,
+    /// The da source channel
+    da_source_channel: Receiver<DaBlockCosts>,
+}
+
+impl<L2, Metadata, DA> GasPriceServiceV1<L2, Metadata, DA>
+where
+    Metadata: MetadataStorage,
+    DA: DaBlockCostsSource,
+{
+    pub fn new(
+        l2_block_source: L2,
+        metadata_storage: Metadata,
+        shared_algo: SharedV1Algorithm,
+        algorithm_updater: AlgorithmUpdaterV1,
+        da_source_adapter_handle: DaSourceService<DA>,
```

Comment:
I feel like we should just be holding onto an abstraction of the da block source here, not the concrete `DaSourceService`. That should make tests easier too.

Reply:
I think @xgreenx suggested we shouldn't do that because we should tightly couple it and not have more traits. I had it initially implemented like that in a previous PR.

Reply:
Hmm. Okay. I might have even been part of that convo. It's not the end of the world.

Reply:
don't understand the action here, do you want me to add the trait or?

Reply:
No. I'm saying we can leave it for now.

Reply:
sounds good